modules/distributor/forwarder/manager.go

```diff
@@ -0,0 +1,370 @@
+func (m *Manager) readOverrides() map[string][]string {
+	result := make(map[string][]string)
+
+	for _, tenantID := range m.overrides.TenantIDs() {
+		names := m.overrides.Forwarders(tenantID)
+		if len(names) < 1 {
+			continue
+		}
+
+		result[tenantID] = names
+	}
+
+	return result
+}
+
+func (m *Manager) updateTenantToForwarderList(tenantToForwarderNames map[string][]string) {
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	oldTenantIDs := make([]string, 0, len(m.tenantToQueueList))
+	for oldTenantID := range m.tenantToQueueList {
+		oldTenantIDs = append(oldTenantIDs, oldTenantID)
+	}
+
+	newTenantIDs := make([]string, 0, len(tenantToForwarderNames))
+	for newTenantID := range tenantToForwarderNames {
+		newTenantIDs = append(newTenantIDs, newTenantID)
+	}
+
+	diff := diff(oldTenantIDs, newTenantIDs)
+
+	for _, removedTenantID := range diff.removed {
+		m.removeQueueListForTenantUnderLock(removedTenantID)
+	}
+
+	for _, addedTenantID := range diff.added {
+		m.updateQueueListForTenantUnderLock(addedTenantID, tenantToForwarderNames[addedTenantID])
+	}
+
+	// Go through all tenants and try to update them.
+	// If forwarder names for particular tenant changed, this will update them.
+	// If names didn't change, this is a noop.
+	for tenantID := range m.tenantToQueueList {
+		m.updateQueueListForTenantUnderLock(tenantID, tenantToForwarderNames[tenantID])
+	}
+}
+
+func (m *Manager) removeQueueListForTenantUnderLock(tenantID string) {
+	ql, found := m.tenantToQueueList[tenantID]
+	if !found {
+		_ = level.Warn(m.logger).Log("msg", "queue list not found", "tenantID", tenantID)
+		return
+	}
+
+	if err := ql.shutdown(context.TODO()); err != nil {
+		_ = level.Warn(m.logger).Log("msg", "failed to shutdown queue list", "tenantID", tenantID)
+	}
+
+	delete(m.tenantToQueueList, tenantID)
+}
+
+// updateQueueListForTenantUnderLock should only be called under tenantToQueueListMu write lock.
+func (m *Manager) updateQueueListForTenantUnderLock(tenantID string, forwarderNames []string) {
+	queueList, found := m.tenantToQueueList[tenantID]
+	if !found {
+		queueList = newQueueList(m.logger, m.reg, tenantID)
+		m.tenantToQueueList[tenantID] = queueList
+	}
+
+	queueList.update(forwarderNames, m.forwarderNameToForwarder)
+}
+
+func (m *Manager) getQueueList(tenantID string) (*queueList, bool) {
+	m.tenantToQueueListMu.RLock()
+	defer m.tenantToQueueListMu.RUnlock()
+
+	queueList, found := m.tenantToQueueList[tenantID]
+
+	return queueList, found
+}
+
+func (m *Manager) shutdown() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	var errs []error
+	for tenantID, ql := range m.tenantToQueueList {
+		if err := ql.shutdown(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("failed to shutdown queuelist for tenantID=%s: %w", tenantID, err))
+		}
+
+		delete(m.tenantToQueueList, tenantID)
```

Comment:
instead of individually deleting create a new, empty map after the loop?

Reply:
Done